scipy/sparse/_coo.py

```diff
@@ -556,31 +572,90 @@ def _add_dense(self, other):
+    def _add_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._add_sparse(self, other)
 
-        if self.ndim == 2:
-            col = self.col
-            row = self.row
-        elif self.ndim == 1:
-            col = self.coords[0]
-            row = np.zeros_like(col)
-        else:
-            raise NotImplementedError(
-                f"coo_matvec not implemented for ndim={self.ndim}")
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = self.__class__((new_data, new_coords), shape=self.shape)         
+        return A
+
+    
+    def _sub_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._sub_sparse(self, other)
+
+        if other.shape != self.shape:
+            raise ValueError(f'Incompatible shapes ({self.shape} and {other.shape})')
+        other = self.__class__(other)
+        new_data = np.concatenate((self.data, -other.data))
+        new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
+        A = coo_array((new_data, new_coords), shape=self.shape)
+        return A
+    
+
+    def _matmul_vector(self, other):
+        if self.ndim < 3:
+            result_shape = self.shape[0] if self.ndim > 1 else 1
+            result = np.zeros(result_shape,
+                              dtype=upcast_char(self.dtype.char, other.dtype.char))
+            if self.ndim == 2:
+                col = self.col
+                row = self.row
+            elif self.ndim == 1:
+                col = self.coords[0]
+                row = np.zeros_like(col)
+            else:
+                raise NotImplementedError(
+                    f"coo_matvec not implemented for ndim={self.ndim}")
+
+            coo_matvec(self.nnz, row, col, self.data, other, result)
+            # Array semantics return a scalar here, not a single-element array.
+            if isinstance(self, sparray) and result_shape == 1:
+                return result[0]
+            return result
+        
+        else: # if dim >= 3
+            result = np.zeros(math.prod(self.shape[:-1]),
+                              dtype=upcast_char(self.dtype.char, other.dtype.char))
+            shape = np.array(self.shape)
+            coords = np.concatenate(self.coords)
+            coo_matvec_nd(self.nnz, len(self.shape), shape, coords, self.data,
+                          other, result)
+            
+            result = result.reshape(self.shape[:-1])
+            return result
+
+
+    def _matmul_dispatch(self, other):
+        if self.ndim < 3:
+            return _data_matrix._matmul_dispatch(self, other)
+        
+        M, N = self.shape[-2], self.shape[-1]
```

Comment:
linting CI test shows M not used any more.
```suggestion
        N = self.shape[-1]
```